components/support/ffi/src/lib.rs

```diff
@@ -343,3 +343,19 @@ fn init_backtraces_once() {
 
 #[cfg(not(feature = "log_backtraces"))]
 fn init_backtraces_once() {}
+
+#[repr(C)]
+pub struct ProtoWrapper {
+    pub data: *mut u8,
+    pub len: i64, // TODO: Kotlin can't @JVMField ULongs.
+}
+
+impl ProtoWrapper {
+    // Can't impl Drop since we pass ProtoWrapper by value.
+    pub fn destroy(self) {
+        let buf = unsafe { std::slice::from_raw_parts_mut(self.data, self.len as usize) };
+        unsafe {
+            drop(Box::from_raw(buf.as_mut_ptr()));
```

Comment:
After your messages in IRC I realized I had meant to take a closer look here.

This violates the guarantees of the allocator interface (which wants us to provide the proper size when deallocating). It should work for malloc/free, but not necessarily for other allocators.

What you should do instead is `drop(Vec::from_raw_parts(self.data, self.len as usize, self.len as usize))`. This isn't documented anywhere, but [apparently was a design constraint for Vec](https://www.reddit.com/r/rust/comments/3ri658/build_boxt_from_raw_parts/cwpc6n4/). Not ideal, but ehh... It's better than what we have, where we know we're passing the wrong size back.